core: condition variable implementation

samkumar opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

This is a simple condition variable implementation. I have found it useful in my programming, so I thought I should contribute it upstream.


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

None at this time.

samkumar commented:
@gebart and @kaspar030 worked on the mutex implementation, so they might be good reviewers for this.

kaspar030 commented on core/condition.c:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+void cond_signal(condition_t* cond)
```
These two functions are candidates for static inlining within the header.

kaspar030 commented on core/include/condition.h:
```diff
@@ -0,0 +1,93 @@
+ * @param[in] cond Condition variable to broadcast.
+ */
+void cond_broadcast(condition_t* cond);
+
```
remove extra newline

samkumar commented:
@kaspar030, thank you for your feedback. I believe I have addressed all of your comments (see commit 806d4e5). I am happy to squash the commits if everything is OK. Please take another look when you get a chance.

samkumar commented:
Actually I decided to go ahead and squash the commits right now so that it's easier to review.

kaspar030 commented:
> Actually I decided to go ahead and squash the commits right now so that it's easier to review.

Next time, please don't. :) We usually squash just before merging so reviewers can keep track of any changes made.

samkumar commented on core/include/cond.h:
```diff
@@ -0,0 +1,222 @@
+#ifndef COND_H
+#define COND_H
+
+#include <cond.h>
```
Honestly I'm not sure how that line got here. It was supposed to be "#include <stdbool.h>". My mistake.

samkumar commented:
@kaspar030, thank you for your second round of feedback. I have fixed the pointer whitespace problems and ran uncrustify on core/cond.c and core/include/cond.h.

I also changed the new constant in thread.h to be "STATUS_COND_BLOCKED" to be consistent with the new name "cond_t".

samkumar commented:
Any update? @kaspar030

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+
+
+/**
+ * @brief Wakes up one thread waiting on the condition variable. The thread is
```
this isn't very brief anymore, I suggest to add an empty line after the first sentence. So that the longer description will go into the detailed section, same applies below.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+
+/**
+ * @brief Wake up threads waiting on a condition variable.
+ * @details For commit purposes you should probably use cond_signal() and
```
please add empty line after `@brief`, and I think you can omit the `@details` then.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+
+/**
+ * @brief Initializes a condition variable.
+ * @details For initialization of variables use COND_INIT instead.
```
same here, add empty line after `@brief` see also below

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+ * true, making the sleeping thread runnable, but the condition may become
+ * false again before the thread is scheduled. To handle this case, the
+ * condition variable should be used in a while loop as follows:
+ *
```
please add markdown style code environment around snippets, here and below.

samkumar replied:
Fixed.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+} cond_t;
+
+/**
+ * @brief   Static initializer for cond_t.
```
add empty line after `@brief`

samkumar replied:
I've added an empty line after `@brief` in all comments.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+
+/**
+ * @brief   Static initializer for cond_t.
+ * @details This initializer is preferable to cond_init().
```
maybe `@note` is more appropriate here?

samkumar replied:
Fixed. I was originally trying to be consistent with the comment for `mutex_init` in mutex.h.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+ * @brief Initializes a condition variable.
+ * @details For initialization of variables use COND_INIT instead.
+ *          Only use the function call for dynamically allocated condition
+ *          variables.
```
add empty line before `@param`

samkumar replied:
I've made sure that there is an empty line before `@param` in all comments.

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,223 @@
+ * @details For initialization of variables use COND_INIT instead.
+ *          Only use the function call for dynamically allocated condition
+ *          variables.
+ * @param[out] cond    pre-allocated condition structure, must not be NULL.
```
looks more like an `[in]` param to me, or not? At least the object/memory is not allocated inside that function and passed out, it just modifies attributes of the `cond_t` passed in.

samkumar replied:
Done. I was originally trying to be consistent with the comment for `mutex_init` in mutex.h.

samkumar commented:
I've now made all of the changes requested by @smlng and @kaspar030.
Sorry for the delay in getting this done. I was busy with paper deadlines at the beginning of May and June.

By the way, the original formatting I had used for comments was based on the style in mutex.h. Perhaps the comments in that file need to be similarly reformatted?

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,235 @@
+
+/**
+ * @brief       Condition variable for thread synchronization
+ * @ingroup     core
```
should be `core_sync`, too?

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,235 @@
+ *
+ * @param[in] cond    Pre-allocated condition structure. Must not be NULL.
+ */
+static inline void cond_init(cond_t *cond)
```
API functions should **not** be `static`, right?

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,235 @@
+ * @param[in] broadcast     If true, wakes up all threads. If false, only
+ *                          wakes up the first waiting thread.
+ */
+void _cond_signal(cond_t *cond, bool broadcast);
```
if you avoid `inline` on `cond_signal` and `cond_broadcast` plus implementing them in the .c file you can omit this here and hide this internal function completely, right?

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,235 @@
+ *
+ * @param[in] cond  Condition variable to signal.
+ */
+static inline void cond_signal(cond_t *cond)
```
should not be `static` either, see above

smlng commented on core/include/cond.h:
```diff
@@ -0,0 +1,235 @@
+ *
+ * @param[in] cond  Condition variable to broadcast.
+ */
+static inline void cond_broadcast(cond_t *cond)
```
should not be `static` either, see above

samkumar commented:
@smlng I have revised the code as you suggested. I also added a test application for the condition variable.

Thanks for the feedback, @smlng and @kaspar030. Please let me know if there's anything else I can do to get this code ready for merging.

samkumar commented:
By the way, I am aware that the Travis CI build is failing, but I believe that once the commits in this pull request are squashed, the problems will go away. Please let me know if I should do anything about this.

samkumar commented on core/cond.c:
```diff
@@ -0,0 +1,87 @@
+    mutex_lock(mutex);
+}
+
+void _cond_signal(cond_t *cond, bool broadcast)
```
Fixed

samkumar commented:
@kaspar030 Thanks for the suggestion. I have changed that function to be static.

samkumar commented:
Any updates? @kaspar030

samkumar commented:
Anything else I need to change in this pull request? @kaspar030 @smlng @gebart

smlng commented:
please squash and you might need to update the `BOARD_INSUFFICIENT_MEMORY` list for the test, see Murdock output which boards have to few mem.

smlng commented:
@kaspar030 can you have another look, to me this look (very) good with solid documentation.

kaspar030 commented:
> @kaspar030 can you have another look, to me this look (very) good with solid documentation.

Agreed! ACK.

kaspar030 commented:
@samkumar please squash!

samkumar commented:
@kaspar030 @smlng Thanks for taking another look! I've updated `BOARD_INSUFFICIENT_MEMORY` for the test, and squashed the pull request to a single commit.

smlng commented:
great work, and thanks for be patient! ACK and GO
